packages/regenerator-transform/src/visit.js

```diff
@@ -177,32 +184,118 @@ function getOuterFnExpr(funPath) {
+      return funPath;
     }
+    const parameters = funPath.node.params.map(function(param) {
+      return t.cloneDeep(param);
+    })
+
+    // turn our FunctionDeclaration into a FunctionExpression with a different
+    // name - our our eventual outerFnExpr needs to be identical to the original
+    // FunctionDeclaration name so that external calls work correctly, and we
+    // can't also use that name in our FunctionExpression because it would shadow
+    // the outerFnExpr in our wrapping scope
+    const functionExpression = t.functionExpression(
+      funPath.scope.generateUidIdentifierBasedOnNode(node.id),
+      parameters,
+      t.cloneDeep(funPath.node.body),
+      funPath.node.generator,
+      funPath.node.async
+    );
 
-    let markDecl = getRuntimeMarkDecl(pp);
-    let markedArray = markDecl.declarations[0].id;
-    let funDeclIdArray = markDecl.declarations[0].init.callee.object;
-    t.assertArrayExpression(funDeclIdArray);
+    // in spite of the above, we want the resulting function's name property
+    // property to be the same as the original so we wrap the FunctionExpression
+    // with:
+    //   Object.defineProperty(functionExpression, 'name', { value: outerFnExpr.name }
+    // this way introspection / prototype chains work the way we'd expect
+    const memberExpression = t.memberExpression(
+      t.identifier('Object'),
+      t.identifier('defineProperty')
+    )
 
-    let index = funDeclIdArray.elements.length;
-    funDeclIdArray.elements.push(node.id);
+    // create a new identifier based on our original node's id
+    outerFnExpr = t.identifier(node.id.name)
 
-    return t.memberExpression(
-      markedArray,
-      t.numericLiteral(index),
-      true
-    );
-  }
+    let decl
+
+    // set name on our functionExpression per the above comments
+    decl = t.variableDeclarator(
+      outerFnExpr,
+      t.callExpression(
+        memberExpression, [
+          functionExpression,
+          t.stringLiteral('name'),
+          t.objectExpression([
+            t.objectProperty(
+              t.stringLiteral('value'),
+              t.stringLiteral(outerFnExpr.name)
+            )
+          ])
+        ]
+      )
+    )
+
+    let markDecl = getRuntimeMarkDecl(pp)
+    t.assertVariableDeclaration(markDecl)
+
+    let declarations = markDecl.declarations;
+    // save our current index
+    let index = declarations.length
+
+    // this can change from pass to pass if something else is altering our body
+    // so we have to look it up every time :/
+    let declarationsIndex = pp.node.body.findIndex((p) => {
+      return t.isVariableDeclaration(p) && p.declarations === declarations;
+    })
+    assert(!isNaN(declarationsIndex));
+
+    // now push our new declaration into our list
+    declarations.push(decl)
+
+    // special case - handle node types that take FunctionDeclarations as
+    // arguments. not sure if there are more than these two, i couldn't find any
+    // in the docs?
```

Comment:
I'm hoping these cases become unnecessary if we simplify or remove the `name` logic above. 🤞